From 929c0d6796f7651cccfcdd8bf0ffa346020643fd Mon Sep 17 00:00:00 2001
From: Min He <min.he@intel.com>
Date: Wed, 23 Mar 2016 18:02:15 +0800
Subject: [PATCH 390/403] add an i915 option gen9_pg_wa_enable

In this patch, an i915 option "gen9_pg_wa_enable" is added. This parameter
is used to enable/disable gen9 power gating WA (default is enabled). When
this option is enabled, Render and Media engine power gating will be
disabled.

This patch will help VPG team to debug power gating related issues.

Signed-off-by: Min He <min.he@intel.com>
---
 drivers/gpu/drm/i915/i915_drv.h    |    1 +
 drivers/gpu/drm/i915/i915_params.c |    7 +++++++
 drivers/gpu/drm/i915/intel_pm.c    |    1 +
 3 files changed, 9 insertions(+)

diff --git a/drivers/gpu/drm/i915/i915_drv.h b/drivers/gpu/drm/i915/i915_drv.h
index 16ad7a4..6e538bd 100644
--- a/drivers/gpu/drm/i915/i915_drv.h
+++ b/drivers/gpu/drm/i915/i915_drv.h
@@ -2721,6 +2721,7 @@ struct i915_params {
 	bool verbose_state_checks;
 	bool nuclear_pageflip;
 	int edp_vswing;
+	bool gen9_pg_wa_enable;
 };
 extern struct i915_params i915 __read_mostly;
 
diff --git a/drivers/gpu/drm/i915/i915_params.c b/drivers/gpu/drm/i915/i915_params.c
index f3369c2..1bffe0e 100644
--- a/drivers/gpu/drm/i915/i915_params.c
+++ b/drivers/gpu/drm/i915/i915_params.c
@@ -55,6 +55,7 @@ struct i915_params i915 __read_mostly = {
 	.edp_vswing = 0,
 	.enable_guc_submission = false,
 	.guc_log_level = -1,
+	.gen9_pg_wa_enable = true,
 };
 
 module_param_named(modeset, i915.modeset, int, 0400);
@@ -195,3 +196,9 @@ MODULE_PARM_DESC(enable_guc_submission, "Enable GuC submission (default:false)")
 module_param_named(guc_log_level, i915.guc_log_level, int, 0400);
 MODULE_PARM_DESC(guc_log_level,
 	"GuC firmware logging level (-1:disabled (default), 0-3:enabled)");
+
+module_param_named(gen9_pg_wa_enable, i915.gen9_pg_wa_enable, bool, 0600);
+MODULE_PARM_DESC(gen9_pg_wa_enable,
+	"Enable gen9 power gating WA (1=enabled [default], 0=disabled)."
+	"When power gating WA enabled, Render and Media engine will not"
+	"be power gated when RC6");
diff --git a/drivers/gpu/drm/i915/intel_pm.c b/drivers/gpu/drm/i915/intel_pm.c
index 0566f1b..82613cd 100644
--- a/drivers/gpu/drm/i915/intel_pm.c
+++ b/drivers/gpu/drm/i915/intel_pm.c
@@ -4728,6 +4728,7 @@ static void gen9_enable_rc6(struct drm_device *dev)
 	 */
 	if (IS_BXT_REVID(dev, 0, BXT_REVID_A1) ||
 	    ((IS_SKL_GT3(dev) || IS_SKL_GT4(dev)) &&
+	     i915.gen9_pg_wa_enable &&
 	     IS_SKL_REVID(dev, 0, REVID_FOREVER)))
 		I915_WRITE(GEN9_PG_ENABLE, 0);
 	else
-- 
1.7.10.4

